test/opt/test_unify_const.cpp

```diff
@@ -0,0 +1,880 @@
+    expected_builder_.AppendTypesConstantsGlobals(CommonTypes());
+  }
+
+  // Checks the validity of the optimization result.
+  void Check() {
+    // unoptimized code
+    const std::string original_before_strip = test_builder_.GetCode();
+    std::string original_without_opnames;
+    std::unordered_set<std::string> original_opnames;
+    std::tie(original_without_opnames, original_opnames) =
+        StripOpNameInstructionsToSet(original_before_strip);
+
+    // expected code
+    std::string expected_without_opnames;
+    std::unordered_set<std::string> expected_opnames;
+    std::tie(expected_without_opnames, expected_opnames) =
+        StripOpNameInstructionsToSet(expected_builder_.GetCode());
+
+    // optimized code
+    std::string optimized_before_strip;
+    bool modified = false;
+    std::tie(optimized_before_strip, modified) =
+        this->template SinglePassRunAndDisassemble<opt::UnifyConstantPass>(
+            test_builder_.GetCode(),
+            /* skip_nop = */ true);
+    std::string optimized_without_opnames;
+    std::unordered_set<std::string> optimized_opnames;
+    std::tie(optimized_without_opnames, optimized_opnames) =
+        StripOpNameInstructionsToSet(optimized_before_strip);
+
+    // Flag "modified" should be returned correctly.
+    EXPECT_EQ(expected_without_opnames != original_without_opnames, modified);
+    // Code except OpName instructions should be exactly the same.
+    EXPECT_EQ(expected_without_opnames, optimized_without_opnames);
+    // OpName instructions can be in different order, but the content must be
+    // the same.
+    EXPECT_EQ(expected_opnames, optimized_opnames);
+  }
+
+  AssemblyBuilder test_builder_;
+  AssemblyBuilder expected_builder_;
+};
+
+using UnifyFrontEndConstantSingleTest =
+    UnifyConstantTest<PassTest<::testing::Test>>;
+
+TEST_F(UnifyFrontEndConstantSingleTest, Basic) {
+  test_builder_
+      .AppendTypesConstantsGlobals({
+          "%unsigned_1 = OpConstant %uint 1",
+          "%unsigned_1_duplicate = OpConstant %uint 1",  // duplicated constant
+      })
+      .AppendInMain({
+          "%uint_var = OpVariable %_pf_uint Function",
+          "OpStore %uint_var %unsigned_1_duplicate",
+      });
+
+  expected_builder_
+      .AppendTypesConstantsGlobals({
+          "%unsigned_1 = OpConstant %uint 1",
+      })
+      .AppendInMain({
+          "%uint_var = OpVariable %_pf_uint Function",
+          "OpStore %uint_var %unsigned_1",
+      })
+      .AppendNames({
+          "OpName %unsigned_1 \"unsigned_1_duplicate\"",  // the OpName
+                                                          // instruction of the
+                                                          // removed duplicated
+                                                          // constant won't be
+                                                          // erased.
+      });
+  Check();
+}
+
+TEST_F(UnifyFrontEndConstantSingleTest, SkipWhenResultIdHasDecorations) {
```

Comment:
Updated this part with two tests in e9d61bc.
1) float w/o decoration
2) struct w/o member decoration